### PR TITLE
feat: spanner emulator support

### PR DESCRIPTION
**Pull request by skuruppu**
- Allows users to set EmulatorDetection through SpannerConnectionStringBuilder.
- Configures integration tests with EmulatorDetection.EmulatorOrProduction.

### CODE REVIEW

**Comment by skuruppu** [comment]
Currently this only addresses the second dot point of the [comment](https://github.com/googleapis/google-cloud-dotnet/pull/4539#issuecomment-597524619). I am still figuring out how to address the third point.

@jskeet I just wanted some quick feedback from you on whether I'm on the right track :) No need to do a detailed review for now.

File: apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/DatabaseAdminClient.g.cs [hunk 1]
```diff
@@ -412,15 +416,15 @@ public sealed partial class DatabaseAdminClientBuilder : gaxgrpc::ClientBuilderB
         public override DatabaseAdminClient Build()
         {
             DatabaseAdminClient client = null;
-            InterceptBuild(ref client);
+            PartialBuild(ref client);
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/DatabaseAdminClient.g.cs:419]: The generated files shouldn't need to be changed. Just keep the method name as `InterceptBuild` and `InterceptBuildAsync`.

File: apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/DatabaseAdminClientBuilder.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+        partial void PartialBuild(ref DatabaseAdminClient client) => client = MaybeCreateEmulatorClientBuilder()?.Build();
+
+        partial void PartialBuildAsync(CancellationToken cancellationToken, ref Task<DatabaseAdminClient> task) =>
+            task = MaybeCreateEmulatorClientBuilder()?.BuildAsync(cancellationToken);
+
+        private DatabaseAdminClientBuilder MaybeCreateEmulatorClientBuilder()
+        {
+            var emulatorEnvironment = GetEmulatorEnvironment(s_emulatorEnvironmentVariables, s_emulatorEnvironmentVariables);
+            return emulatorEnvironment is null ? null :
+                new DatabaseAdminClientBuilder
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/DatabaseAdminClientBuilder.cs:49]: It may be worth adding a comment to the effect that you're not copying the EmulatorDetection property here, deliberately - that's what keeps it from recursing infinitely.

File: apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/DatabaseAdminClient.g.cs (hunk 1 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/DatabaseAdminClient.g.cs:419]: Ok cool, I think you also mean to change the name in the builder class. I checked to see if anything else was implementing the Intercept* methods but doesn't seem that way so I think this is safe. It compiles at least.

Comment (jskeet) [reply apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/DatabaseAdminClient.g.cs:419]: Yes, exactly. They were introduced primarily for this feature, so there really shouldn't be anything else conflicting with it :)

File: apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/DatabaseAdminClientBuilder.cs (hunk 2 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.Admin.Database.V1/Google.Cloud.Spanner.Admin.Database.V1/DatabaseAdminClientBuilder.cs:51]: I added a comment to each builder. Hope this is ok.

**Comment by skuruppu** [comment]
This is ready for review (I do need to fix the unit tests which I'll do now). I ran the ITs and they ran for the most part:

Total tests: 167
     Passed: 92
     Failed: 74
    Skipped: 1

I will go through the failed tests to figure out which ones need to be skipped (due to the emulator not supporting all features that the production backend supports). This will need to be done in a separate PR. If there's anything that can be fixed in terms of the setup, I will do so in this PR.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs [hunk 3]
```diff
@@ -69,6 +69,16 @@ static SessionPoolManager()
...
 
+        /// <summary>
+        /// Specifies whether to allow the connection to check for the presence of the emulator
+        /// environment variable.
+        /// </summary>
+        /// <remarks>
+        /// This property defaults to <see cref="EmulatorDetection.None"/>, meaning that the
+        /// environment variable is ignored.
+        /// </remarks>
+        public EmulatorDetection EmulatorDetection { get; set; }
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:80]: I don't think this belongs here. It's generally immutable.

I don't *think* you need to change SessionPoolManager at all - because the emulator detection should instead affect the SpannerClientCreationOptions that are in use.
SpannerClientCreationOptions gets its information from a SpannerConnectionStringBuilder - so it can use the settings there.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs [hunk 4]
```diff
@@ -313,7 +323,8 @@ private static async Task<SpannerClient> CreateClientAsync(SpannerClientCreation
             return new SpannerClientBuilder
             {
                 CallInvoker = callInvoker,
-                Settings = spannerSettings
+                Settings = spannerSettings,
+                EmulatorDetection = emulatorDetection
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:327]: Setting a call invoker and emulator detection would be an odd thing to do. (I'm slightly surprised we don't prohibit it.)
If we do the work in SpannerClientCreationOptions (which may mean some code duplication, unfortunately - let's see what we can do) the endpoint and credentials will already be set correctly, so we shouldn't need to make any changes here.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs [hunk 5]
```diff
@@ -307,6 +307,20 @@ public SessionPoolManager SessionPoolManager
...
 
+        /// <summary>
+        /// Specifies whether to allow the connection to check for the presence of the emulator
+        /// environment variable.
+        /// </summary>
+        /// <remarks>
+        /// This property defaults to <see cref="EmulatorDetection.None"/>, meaning that the
+        /// environment variable is ignored.
+        /// </remarks>
+        public EmulatorDetection EmulatorDetection
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:318]: I think this should be handled in a fashion much more like properties such as MaximumGrpcChannels, so it becomes part of the connection string itself.

**Comment by skuruppu** [comment]
Thanks @jskeet. I agree about doing this as two separate PRs. To preserve the discussion, I may leave this PR as is to continue the work on setting the connection properties. I'll cherry-pick the builders into a separate PR.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs (hunk 4 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:327]: Oh I see, so we shouldn't rely on the automated emulator detection for the data client. We should only do that for the admin clients. So we'll manually read the env var and set the endpoint and credentials. Is the worry that it would confuse users because they would set an endpoint and credentials but we'd be using a different parameters for the actual connection?

Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:327]: It's not a matter of "not relying" on the automated emulation code - it's not that we don't trust it. It's a matter of making it fit in well with the rest of the infrastructure.
I'd like to investigate how to do this without code duplication though - for example, it may be possible to add a *public* member to `SpannerClientBuilder` to say "please let me know how you would be configured for emulation" (effectively like the current partial method does) and then use the endpoint and credentials from that.

I think our aim should be for the behavior to be consistent between "using SpannerClientBuilder directly" and "using SpannerConnectionStringBuilder directly" in terms of the connection.

**Comment by jskeet** [comment]
I've merged #4928 now, so it's probably worth rebasing this PR, unless you'd prefer to start a new one.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs (hunk 5 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:318]: All done.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs (hunk 3 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs:80]: Ok that makes sense. Now we're getting the value from SpannerClientCreationOptions through SpannerConnectionStringBuilder.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs (hunk 4 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SessionPoolManager.cs]: I've made the minimal amount of changes to get this working. But I'm still not updating the endpoint and credentials in the connection string to reflect what is actually being connected to. I think this is where your comment regarding updating the `SpannerClientBuilder` comes in.

The other issue that I want to solve is that if the connection string has EmulatorDetection then users shouldn't be asked to specify credentials. I'll also work on trying to bypass this check.

**Comment by jskeet** [comment]
I'll get to this when I can, but it may go into next week I'm afraid.

**Comment by skuruppu** [comment]
No worries @jskeet. I'm also sporadically working on it when I find bits of time here and there.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs [hunk 6]
```diff
@@ -289,6 +290,31 @@ public int Timeout
...
+                if (TryGetValue(EmulatorDetectionKeyword, out object value) &&
+                    (value is EmulatorDetection parsed || (value is string text && EmulatorDetection.TryParse(text, out parsed))))
+                {
+                    return parsed >= EmulatorDetection.None && parsed <= EmulatorDetection.EmulatorOrProduction ? parsed : EmulatorDetection.None;
+                }
+                return EmulatorDetection.None;
+            }
+            set
+            {
+                this[EmulatorDetectionKeyword] = (int) value;
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerConnectionStringBuilder.cs:314]: Setting it to the integer value will make the connection string very hard to read in string form. I would call ToString - after validating that it's a valid value using `GaxPreconditions.CheckEnumValue`.

**Comment by skuruppu** [comment]
This is what I came up with for now https://github.com/googleapis/google-cloud-dotnet/pull/4915/commits/e98b470ec20d5318fc13040a82c6a104e737711a. I don't think you need to move the CallInvoker to SpannerClientCreationOptions, as long as the endpoint and credentials can be correctly extracted from that class.

My proposed change to SpannerClientBuilder is a bit of a mess since there's no nice way to extract the endpoint and credentials without attempting to build an instance of the class. For now I'm reading the env var manually. But if there was a static version of `GetEmulatorEnvironment` that also accepts an `EmulatorDetection` argument, then it might be cleaner.

Sorry, I know the back-and-forth is taking a bit of time. I'm more than happy for you to take a crack at this but I know you have a lot more stuff going on so I want to help as much as possible.

**Comment by jskeet** [comment]
Building an instance of `SpannerClientBuilder` is the way to go - we want to get the same exact behaviour as if we were specifying the emulator detection in the builder.

I would suggest just making `MaybeCreateEmulatorClientBuilder` public - then you can use that in `SessionPoolManager`. If it returns a non-null value, use the endpoint and credentials from that.

**Comment by skuruppu** [comment]
PTAL @jskeet. I made all the discussed changes.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerClientCreationOptions.cs [hunk 7]
```diff
@@ -120,6 +135,19 @@ public override string ToString()
...
+            var builder = new SpannerClientBuilder
+            {
+                EmulatorDetection = EmulatorDetection
+            };
+            var emulatorBuilder = builder.MaybeCreateEmulatorClientBuilder();
+            if (emulatorBuilder != null)
+            {
+                return emulatorBuilder.ChannelCredentials;
+            }
+
```
Comment (skuruppu) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerClientCreationOptions.cs:150]: I was a bit surprised I had to do this.

I was attempting to call this function conditionally in `SessionPoolManager.CreateClientAsync()` only if `MaybeCreateEmulatorClientBuilder()` returns null. But this call kept happening and returning an error saying that a credentials files is not given  so I assume that it has something to do with the fact that it's async. I was trying to read up on when async functions get called but didn't find a good source that explained my problem.

**Comment by jskeet** [comment]
Will have a look when I get a chance, but it may well be quite late in the week.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientBuilder.cs [hunk 8]
```diff
@@ -42,7 +42,16 @@ public new EmulatorDetection EmulatorDetection
         partial void InterceptBuildAsync(CancellationToken cancellationToken, ref Task<SpannerClient> task) =>
             task = MaybeCreateEmulatorClientBuilder()?.BuildAsync(cancellationToken);
 
-        private SpannerClientBuilder MaybeCreateEmulatorClientBuilder()
+        /// <summary>
+        /// May return a builder that will connect to the emulator under certain conditions.
+        /// </summary>
+        /// <remarks>
+        /// Returns null if <see cref="EmulatorDetection"/> is <see cref="EmulatorDetection.None"/>
+        /// or <see cref="EmulatorDetection.ProductionOnly"/>. Otherwise, returns a builder to
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientBuilder.cs:50]: I'd write this as "it returns null if the emulator should not be used" - it will return null if EmulatorDetection is EmulatorOrProduction but the environment variable isn't configured, for example.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientBuilder.cs [hunk 9]
```diff
@@ -42,7 +42,16 @@ public new EmulatorDetection EmulatorDetection
...
+        /// connect to the emulator using the endpoint specified by the emulator host environment
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientBuilder.cs:51]: I'd probably say it returns a builder with the endpoint and credentials set appropriately for the emulator.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerClientCreationOptions.cs (hunk 7 above)
Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerClientCreationOptions.cs:150]: I can't tell what the code looked like when it was failing, but it seems reasonable to me. I don't think this has anything to do with async.

Basically if we want to use the emulator, we need to specify insecure channel credentials - otherwise, we use the normal code.

I'd probably write it in one step though:

```csharp
var emulatorBuilder = new SpannerClientBuilder { EmulatorDetection = EmulatorDetection }.MaybeCreateEmulatorClientBuilder();
...
```

If we use it elsewhere for the endpoint, you might want to put this into a common method.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabase.cs [hunk 10]
```diff
@@ -79,25 +81,50 @@ private SpannerTestDatabase(string projectId)
...
+            // emulator (if configured) so that we can configure the session
+            // pool. This is necessary because the emulator doesn't support
+            // running concurrent transactions so we need to configure the
+            // session pool to only have one active session at a time so that
+            // we don't prepare multiple concurrent transactions.
+            var clientBuilder = new SpannerClientBuilder
+            {
+                EmulatorDetection = emulatorDetection
             };
+            var emulatorBuilder = clientBuilder.MaybeCreateEmulatorClientBuilder();
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabase.cs:106]: I wonder whether this should be in the production code, if everyone will always need to do that.
Do we know whether this is a long-term restriction?

Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabase.cs:106]: In general, the changes in this file (beyond just setting EmulatorDetection) feel like they're a symptom of us not doing quite enough to "just make it work".

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabase.cs [hunk 11]
```diff
@@ -79,25 +81,50 @@ private SpannerTestDatabase(string projectId)
...
             {
                 builder.Port = int.Parse(SpannerPort);
             }
             NoDbConnectionString = builder.ConnectionString;
             SpannerClientCreationOptions = new SpannerClientCreationOptions(builder);
-            var databaseBuilder = builder.WithDatabase(SpannerDatabase);
-            ConnectionString = databaseBuilder.ConnectionString;
-            DatabaseName = databaseBuilder.DatabaseName;
+            Builder = builder.WithDatabase(SpannerDatabase);
+            Builder.EmulatorDetection = emulatorDetection;
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabase.cs:121]: We shouldn't need to set the EmulatorDetection again, should we?

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabase.cs [hunk 12]
```diff
@@ -116,6 +143,6 @@ private static string GetEnvironmentVariableOrDefault(string name, string defaul
             return string.IsNullOrEmpty(value) ? defaultValue : value;
         }
         
-        public SpannerConnection GetConnection() => new SpannerConnection(ConnectionString);        
+        public SpannerConnection GetConnection() => new SpannerConnection(Builder);
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabase.cs:146]: Why is this change required? I'd expect the connection string to be fine.

Comment (jskeet) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data.CommonTesting/SpannerTestDatabase.cs:146]: Or is this because we've set the SessionPoolManager because of the options? It feels like a lot of this change could go away if we didn't need to set the options explicitly.

File: apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerClientCreationOptions.cs (hunk 7 above)
Comment (skuruppu) [reply apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerClientCreationOptions.cs:150]: Thanks for the suggestion. Now it's in a separate function and I've cleaned up the calls a bit. Now there's no need to modify SessionPoolManager at all and all the changes are confined to this file.

**Comment by skuruppu** [comment]
I addressed most of the changes except for the ones in https://github.com/googleapis/google-cloud-dotnet/pull/4915/files#diff-450b4cc43d8fb620f621d63b7de37a31 relating to the session pool.

For context, the changes to that file came out of #4970 when I was trying to run the existing ITs against the emulator. To be honest, I haven't quite gotten the whole thing to work yet even with that change.

My preference would be to revert the changes to that file and work on it separately. The highest priority is to add the emulator support to the client since there is a customer that's keen to try this out in C#.

Getting the ITs to work is a bonus so that we can report emulator issues to the dev team and perhaps have a presubmit that runs the ITs against the emulator. But this will require a bunch of work from my side and from the emulator team to fix some of the failing tests.

@jskeet I would be happy for you to try out a fix for the session pool issue. I'm not familiar with it enough to know what to change. But only if you have time. Otherwise, I'll work on it until I figure it out.

I don't know if the emulator team is planning to fix the concurrent transaction issue before GA. I will check with them and get back to you.

File: apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientBuilder.cs [hunk 13]
```diff
@@ -46,10 +46,9 @@ public new EmulatorDetection EmulatorDetection
         /// May return a builder that will connect to the emulator under certain conditions.
         /// </summary>
         /// <remarks>
-        /// Returns null if <see cref="EmulatorDetection"/> is <see cref="EmulatorDetection.None"/>
-        /// or <see cref="EmulatorDetection.ProductionOnly"/>. Otherwise, returns a builder to
-        /// connect to the emulator using the endpoint specified by the emulator host environment
-        /// variable, and insecure credentials.
+        /// It returns null if the emulator should not be used, i.e. if <see cref="EmulatorDetection"/> is
```
Comment (jskeet) [inline apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SpannerClientBuilder.cs:49]: Nit: change the `i.e.` to `e.g.`, as there are other cases where the emulator shouldn't be used. But otherwise, this is fine, thanks.

**Comment by skuruppu** [comment]
All done, thanks for the review feedback @jskeet.
